### PR TITLE
Support Blueprint Permissions

### PR DESCRIPTION
**Pull request by erikzaadi**
# Description

What - Add support for controlling Blueprint Permissions
Why  - Allow a more fine grained control for Blueprint operations
How  - Add new blueprint-permissions package

## Type of change

New feature (non-breaking change which adds functionality)

- [x] Add Blueprint permissions
- [x] Add tests
- [x] Update hcl markdown docs

### CODE REVIEW

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
Nice Job @erikzaadi ! left some comments

File: port/blueprint-permissions/resource.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+func (r *BlueprintPermissionsResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	var state *BlueprintPermissionsModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	// blueprintPermissions is not deletable resource by itself as it is tied to an blueprint and is created by default when an blueprint is created
```
Comment (Tankilevitch) [inline port/blueprint-permissions/resource.go:114]:
```suggestion
	// blueprintPermissions is not a deletable resource by itself, as it is tied to a blueprint and is created by default when a blueprint is created
```

File: port/blueprint-permissions/resource_test.go [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+	var testAccPortBlueprintResourceBasic = createBlueprintWithProperties(blueprintIdentifier)
+
+	teamName := utils.GenID()
+	var testAccBaseBlueprintPermissionsConfigUpdate = fmt.Sprintf(`
+
+	resource "port_team" "team" {
+		name = "%s"
+		description = "Test description"
+		users = []
+	}
```
Comment (Tankilevitch) [inline port/blueprint-permissions/resource_test.go:172]: you are not using this in the test

File: port/blueprint-permissions/resource_test.go [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+`, blueprintEnvIdentifier, blueprintMicroserviceIdentifier)
+
+	teamName := utils.GenID()
+	var testAccBaseBlueprintPermissionsConfigUpdate = fmt.Sprintf(`
+
+	resource "port_team" "team" {
+		name = "%s"
+		description = "Test description"
+		users = []
+	}
```
Comment (Tankilevitch) [inline port/blueprint-permissions/resource_test.go:318]: same here, though I would like to have at least one test checking that the teams field works

File: port/blueprint-permissions/schema.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+						},
+					},
+				},
+				"update_properties": schema.MapNestedAttribute{
+					Optional: true,
+					NestedObject: schema.NestedAttributeObject{
+						Attributes: getAssigneeProps("update specific property"),
+					},
+					Validators: []validator.Map{
+						mapvalidator.KeysAre(
```
Comment (Tankilevitch) [inline port/blueprint-permissions/schema.go:95]: missing descriptions

File: port/blueprint-permissions/schema.go [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+				"unregister": schema.SingleNestedAttribute{
+					Required:   true,
+					Attributes: getAssigneeProps("unregister"),
+				},
+				"update": schema.SingleNestedAttribute{
+					Required:   true,
+					Attributes: getAssigneeProps("update"),
+				},
+				"update_metadata_properties": schema.SingleNestedAttribute{
+					Required: true,
```
Comment (Tankilevitch) [inline port/blueprint-permissions/schema.go:69]: missing description, also this is a "special" property we added which isn't aligned with the API, so we should add more explanation about it

File: port/blueprint-permissions/schema.go [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+			"Member",
+		],
+		"users": [],
+		"teams": []
+	},
+	"update_metadata_properties" = {
+		"icon" = {
+			"roles": [
+				"Member",
+			],
```
Comment (Tankilevitch) [inline port/blueprint-permissions/schema.go:158]: see where you can add "admin"  role example to one of the "actions" so they'll see the diversity

File: port/blueprint-permissions/blueprintPermissionToPortBody.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+	}
+	var finalUpdateProperties cli.BlueprintRolesOrPropertiesPermissionsBlock = nil
+	if updateMetadataProperties != nil {
+		finalUpdateProperties = updateMetadataProperties
+	}
+	if updateProperties != nil {
+		if finalUpdateProperties == nil {
+			finalUpdateProperties = updateProperties
+		} else {
+			utils.CopyGenericMaps(finalUpdateProperties, updateProperties)
```
[the remark this answers is not in the document]
Comment (erikzaadi) [reply port/blueprint-permissions/blueprintPermissionToPortBody.go:94]: that's basically what copy does, it merges all to the destination map, I could make that immutable and return a new map, then merge would make more sense, WDYT?

File: port/blueprint-permissions/resource_test.go (hunk 3 above)
Comment (erikzaadi) [reply port/blueprint-permissions/resource_test.go:311]: Added

File: port/blueprint-permissions/schema.go (hunk 5 above)
Comment (erikzaadi) [reply port/blueprint-permissions/schema.go:69]: Added

File: port/blueprint-permissions/schema.go (hunk 6 above)
Comment (erikzaadi) [reply port/blueprint-permissions/schema.go]: Done

File: port/blueprint-permissions/schema.go [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+var BlueprintPermissionsResourceMarkdownDescription = `
+
+# Blueprint Permissions resource
+
+## Example Usage
+
+### Allow access to all members:
+
+` + "```hcl" + `
+resource "port_blueprint_permissions" "microservices_permissions" {
```
[the remark this answers is not in the document]
Comment (erikzaadi) [reply port/blueprint-permissions/schema.go]: Done

File: port/blueprint-permissions/schema.go (hunk 4 above)
Comment (erikzaadi) [reply port/blueprint-permissions/schema.go]: Added

File: examples/resources/port_blueprint_permissions/main.tf [hunk 9]
```diff
@@ -1,3 +1,18 @@
...
+    }
+  }
+}
+provider "port" {
+  client_id = "60EsooJtOqimlekxrNh7nfr2iOgTcyLZ"     # or set the environment variable PORT_CLIENT_ID
+  secret    = "UU3HcDOl2tvv8Y71FClcfMZmxEDP3Unz73WWaD8ODsrrGhLkFi839K0VaqMGsIrA" # or set the environment variable PORT_CLIENT_SECRET
+  base_url  = "http://localhost:3000"
+}
+
+
```
Comment (Tankilevitch) [inline examples/resources/port_blueprint_permissions/main.tf:15]: oopsi? at least its local permissions 😮‍💨

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

File: examples/resources/port_blueprint_permissions/main.tf (hunk 9 above)
Comment (erikzaadi) [reply examples/resources/port_blueprint_permissions/main.tf:15]: ZOMG

**Review by Tankilevitch — APPROVED** [review]
Looks good 🥇

**Pull request merged**